src/kernel/src/dmem/mod.rs

```diff
@@ -1,40 +1,96 @@
+use thiserror::Error;
+
+use crate::dev::{Dmem, DmemContainer};
 use crate::errno::EINVAL;
-use crate::fs::Fs;
-use crate::process::VThread;
+use crate::fs::{make_dev, CharacterDevice, DriverFlags, Fs, MakeDevError, MakeDevFlags, Mode};
+use crate::info;
+use crate::process::{VProc, VThread};
 use crate::syscalls::{SysErr, SysIn, SysOut, Syscalls};
+use crate::ucred::{Gid, Privilege, Uid};
 use std::sync::Arc;
 
 pub use self::blockpool::*;
 
 mod blockpool;
 
+#[derive(Debug)]
+struct DmemDevice {
+    name: &'static str,
+    dev: Arc<CharacterDevice>,
+}
+
+impl DmemDevice {
+    pub const NAMES: [&'static str; 3] = ["dmem0", "dmem1", "dmem2"];
+
+    pub(super) fn new(name: &'static str, dev: Arc<CharacterDevice>) -> Self {
+        Self { name, dev }
+    }
+}
+
 /// An implementation of direct memory system on the PS4.
 pub struct DmemManager {
     fs: Arc<Fs>,
+    dmem: [DmemDevice; 3],
 }
 
 impl DmemManager {
+    const DMEM_TOTAL_SIZE: usize = 0x13C_000_000;
+
     pub fn new(fs: &Arc<Fs>, sys: &mut Syscalls) -> Arc<Self> {
-        let dmem = Arc::new(Self { fs: fs.clone() });
+        let dmem_devices: [DmemDevice; 3] = [0 as usize, 1, 2]
```

Comment:
You could do an array of DmemContainer directly, the get the index by calling .enumerate(),

Reply:
Although, I'm not sure if that can be turned back into an array. Maybe it can't really be created in a loop.

Reply:
It can't :c

Reply:
Yeah, I looked it up, this doesn't use iterators :(

Reply:
I would just consider making each separately and having them as fields separately rather than an array.